src/Drupal/Commands/core/MigrateRunnerCommands.php

```diff
@@ -0,0 +1,694 @@
+            throw new \Exception(dt('No migrations found.'));
+        }
+
+        $userData = [
+            'options' => array_intersect_key($options, array_flip([
+                'limit',
+                'feedback',
+                'idlist',
+                'update',
+                'force',
+                'timestamp',
+                'total',
+                'progress',
+            ])),
+            'execute_dependencies' => $options['execute-dependencies'],
+        ];
+
+
+        // Include the file providing a 'migrate_prepare_row' hook implementation.
+        require_once Path::join(DRUSH_BASE_PATH, 'includes/migrate_runner.inc');
+
+        // Take it one group at a time, importing the migrations within each group.
+        foreach ($list as $tag => $migrations) {
+            array_walk($migrations, [static::class, 'executeMigration'], $userData);
+        }
+    }
+
+    /**
+     * Executes a single migration.
+     *
+     * If the --execute-dependencies option was given, the migration's
+     * dependencies will also be executed first.
+     *
+     * @param \Drupal\migrate\Plugin\MigrationInterface $migration
+     *   The migration to execute.
+     * @param string $migrationId
+     *   The migration ID (not used, just an artifact of array_walk()).
+     * @param array $userData
+     *   Additional data passed to the callback.
+     *
+     * @throws \Exception
+     *   If there are failed migrations.
+     */
+    protected function executeMigration(MigrationInterface $migration, string $migrationId, array $userData): void
+    {
+        static $executedMigrations = [];
+
+        if ($userData['execute_dependencies']) {
+            $dependencies = $migration->getMigrationDependencies()['required'];
+            // Remove already executed migrations.
+            $dependencies = array_diff($dependencies, $executedMigrations);
+            if ($dependencies) {
+                $requiredMigrations = $this->getMigrationPluginManager()->createInstances($dependencies);
+                array_walk($requiredMigrations, [static::class, __FUNCTION__], $userData);
+            }
+        }
+        if (!empty($userData['options']['force'])) {
+            $migration->set('requirements', []);
+        }
+        if (!empty($userData['options']['update'])) {
+            if (empty($userData['options']['idlist'])) {
+                $migration->getIdMap()->prepareUpdate();
+            } else {
+                $sourceIdValuesList = MigrateUtils::parseIdList($userData['options']['idlist']);
+                $keys = array_keys($migration->getSourcePlugin()->getIds());
+                foreach ($sourceIdValuesList as $sourceIdValues) {
+                    $migration->getIdMap()->setUpdate(array_combine($keys, $sourceIdValues));
+                }
+            }
+        }
+
+        // Initialize the Symfony Console progress bar, if case.
+        $this->initProgressBar($migration, $userData['options']);
```

Comment:
I didnt yet fully investigate this PR to see how we are doing progress. I just want o point out that we added one in #4583 and it used `$this->io()` which I dont see here.

Reply:
@weitzman unfortunately, for testing purposes, I need to pass the 3rd argument to `\Symfony\Component\Console\Helper\ProgressBar::__construct()` (>= Symfony 4). Otherwise I cannot predict the progress bar output, thus I cannot properly test. But `$this->io->progressStart()` doesn't allow to pass that argument.

Reply:
@weitzman I have 2 options:

- Keep the current approach
- Switch to `$this->io->progressStart()` but dramatically simplify that test.

Reply:
Lets keep the current approach and keep the testing. Thanks.